lib/schema_checker.py

```diff
@@ -0,0 +1,118 @@
+import os
+import string
+import re
+from schema import Schema, SchemaError, Optional, Or, Use
+# https://docs.green-coding.berlin/docs/measuring/usage-scenario/
+# networks documentation is different than what i see in the wild!
+    # name: str
+    # also isn't networks optional?
+    # fix documentation - name not needed, also netowrks optional
+    # add check in runner.py networks parsing, make sure its valid_string
+    # is services/type optional?
+
+# services /type missing from documentation?
+
+
+# https://github.com/compose-spec/compose-spec/blob/master/spec.md
+
+def single_or_list(value):
+    return Or(value, [value])
+
+def is_valid_string(value):
+    valid_chars = set(string.ascii_letters + string.digits + '_' + '-')
+    if not set(value).issubset(valid_chars):
+        raise SchemaError(f"{value} does not use valid characters! (a-zA-Z0-9_-)")
+
+def contains_no_invalid_chars(value):
+    bad_values = re.findall(r'(\.\.|\$|\'|"|`|!)', value)
+    if bad_values:
+        raise SchemaError(f"{value} includes disallowed values: {bad_values}")
+
+## If we ever want smarter validation for ports, here's a list of examples that we need to support:
+# - 3000
+# - 3000-3005
+# - 8000:8000
+# - 9090-9091:8080-8081
+# - 49100:22
+# - 8000-9000:80
+# - 127.0.0.1:8001:8001
+# - 127.0.0.1:5000-5010:5000-5010
+# - 6060:6060/udp
+
+def validate_compose_include(value):
+    #!include compose.yml
+    # make sure value starts with the string '!include', then space, then a valid yaml file name
+    # make sure that compose yaml exists
+    if not value.startswith('!include '):
+        raise SchemaError(f"{value} does not begin with !include")
+    if (not value.endswith('.yml') and not value.endswith('.yaml')):
+        raise SchemaError(f"{value} does not end with a valid yaml filename")
+
+    # Extract the filename from the value
+    filename = value[len('!include '):]
+
+    # Make sure that the compose yaml exists
+    if not os.path.exists(filename):
```

Comment:
At the moment its optional - if the compose-file line is missing, the validation will still pass (as its not necessary anyways)

Instead of having a flag, I think we can have the validate_compose_include function which validates this line handle the different cases. I think I need a bit more details about what would be different exactly to understand how to best handle this case

Reply:
Here is the ticket: https://github.com/green-coding-berlin/green-metrics-tool/issues/270
We will still have the case that the `!include` line will be present but the file will not be found as we don't check out the whole repo. Only the one file.

Reply:
Updated this - turns out there's no easy way to just pass in a flag through the validator down the chain, so instead I turned the schema checker into a class in order to store the flag as a class variable. That does the trick. You can now set a flag for whether it will actually check if the .yaml file is there or if the line is just syntactically correct